dizoo/gym_anytrading/envs/trading_env.py

```diff
@@ -0,0 +1,231 @@
+
+    def reset(self, start_idx: int = None) -> Any:
+        self.cnt += 1
+        self.prices, self.signal_features = self._process_data(start_idx)
+        self._done = False
+        self._current_tick = self._start_tick
+        self._last_trade_tick = self._current_tick - 1
+        self._position = Positions.FLAT
+        self._position_history = [self._position]
+        self._profit_history = [1.]
+        self._total_reward = 0.
+
+        return self._get_observation()
+
+    def step(self, action: List) -> Any:
+        action = int(action[0])
+
+        self._done = False
+        self._current_tick += 1
+
+        if self._current_tick == self._end_tick:
+            self._done = True
+
+        step_reward = self._calculate_reward(action)
+        self._total_reward += step_reward
+
+        self._position, trade = transform(self._position, action)
+
+        if trade:
+            self._last_trade_tick = self._current_tick
+
+        self._position_history.append(self._position)
+        self._profit_history.append(float(np.exp(self._total_reward)))
+        observation = self._get_observation()
+        info = dict(
+            total_reward=self._total_reward,
+            position=self._position.value,
+        )
+
+        if self._done:
+            if self._env_id[-1] == 'e' and self.cnt % self.plot_freq == 0:
+                self.render()
+            info['max_possible_profit'] = np.log(self.max_possible_profit())
+            info['final_eval_reward'] = self._total_reward
+
+        return BaseEnvTimestep(observation, step_reward, self._done, info)
+
+    def _get_observation(self):
+        obs = to_ndarray(self.signal_features[(self._current_tick - self.window_size + 1):self._current_tick + 1]
+                         ).reshape(-1).astype(np.float32)
+        obs = np.hstack([obs, to_ndarray([self._position.value])]).astype(np.float32)
+
+        return obs
+
+    def render(self) -> None:
+        plt.clf()
+        plt.plot(self._profit_history)
+        plt.savefig(self.save_path + str(self._env_id) + "-profit.png")
+
+        plt.clf()
+        window_ticks = np.arange(len(self._position_history))
+        eps_price = self.raw_prices[self._start_tick:self._end_tick + 1]
+        plt.plot(eps_price)
+
+        short_ticks = []
+        long_ticks = []
+        flat_ticks = []
+        for i, tick in enumerate(window_ticks):
+            if self._position_history[i] == Positions.SHORT:
+                short_ticks.append(tick)
+            elif self._position_history[i] == Positions.LONG:
+                long_ticks.append(tick)
+            else:
+                flat_ticks.append(tick)
+
+        plt.plot(short_ticks, eps_price[short_ticks], 'ro')
+        plt.plot(long_ticks, eps_price[long_ticks], 'go')
+        plt.plot(flat_ticks, eps_price[flat_ticks], 'bo')
+        plt.savefig(self.save_path + str(self._env_id) + '-price.png')
+
+    def close(self):
+        plt.close()
+
+    def _process_data(self):
```

Comment:
you can use `ABC` and `abstractmethod` to restrict that subclasses must implement these three methods